TelemetryDesignDoc.md

```diff
@@ -0,0 +1,327 @@
+
+Shuffleboard is, essentially, stable abandonware.  However, it has a number of known performance issues and complaints about stability, resource use, and general user experience remain frequent year-after-year.  Given the lack of JavaFX developer impetus/availability/expertise on the WPILib team, and the irrelevance of the platform to the current software "zeitgeist," this is a major stumbling block so long as Shuffleboard's core value-add features (user-code-configurable tab/layout telemetry/dashboard viewer) remain non-duplicated within the ecosystem.
+
+### Overloading of the Sendable implementation
+
+On the surface, the `Sendable` interface offers a powerful abstraction to teams who wish to log their state declaratively rather than imperatively.  However, behind-the-scenes `Sendable` pulls double-duty as the backbone of the `LiveWindow` implementation, which is responsible for auto-populating a list of sensors and actuators for easy debugging.
+
+The `LiveWindow` design philosophy holds that all sensor and actuator objects in code should be accessible via the dashboard while the robot is in Test Mode, *regardless of what the user has done in their robot code.*  While this is a nice idea in theory, it places *huge* constraints on the handling of the `Sendable` interface which often interfere with the latter's use as a telemetry handler:
+
+* While writes to the dashboard work as expected, reads from the dashboard (i.e. calls to the bound setters) are suspended *except* in LiveWindow mode.  This is not documented anywhere on the class, and is extremely inconvenient.
+* The API includes an `isActuator` flag which seemingly serves no purpose other than to disable certain widgets on ShuffleBoard (it does not interact with the other dashboards).
+* Sendables can be registered with `LiveWindow` even if they have no business being there (i.e. they are not a sensor or an actuator).
+* It is not documented which `Sendable`s automatically add themselves to the `LiveWindow` registry immediately upon construction, and the application of the `Sendable` interface to certain classes which would otherwise be able to provide useful telemetry to teams (e.g. the various pneumatics classes) has been delayed or abandoned due to the complexities of maintaining a such a static registry, even when the classes could feasibly implement the interface and expose themselves to users sans automatic registration.
+  
+### Crude interaction between Sendable and Commands
+
+Parts of the `Sendable` interface - particulary those involving the `LiveWindow` registry - have hard dependencies on the design philosophy of the Command-based framework.  In particular, `LiveWindow` widgets are grouped by "subsystem," even though they are populated automatically on robots that may not have recognizable `Subsystem` classes.  The relevant values are silently defaulted - and the functionality rendered useless - in all but a very narrow set of circumstances.
```

Comment:
It feels hacky and arbitrary to have a single flat grouping mechanism which is separate from the existing pathstring hierarchy, and which borrows terminology from an unrelated other part of the library.  Since NT4 has metadata, it feels natural to use that to indicate grouping as it doesn't involve setting aside a flag in code for something that very well might not be used at all.

Reply:
So you're keeping the concept of subsystem, but not putting it into the NT path, but rather adding it to the metadata?  Seems reasonable.